src/main/java/com/faforever/client/login/OAuthValuesReceiver.java

```diff
@@ -0,0 +1,155 @@
+package com.faforever.client.login;
+
+import com.faforever.client.config.ClientProperties;
+import com.faforever.client.fx.JavaFxUtil;
+import com.faforever.client.fx.PlatformService;
+import com.faforever.client.i18n.I18n;
+import com.faforever.client.update.ClientConfiguration.OAuthEndpoint;
+import com.faforever.client.user.UserService;
+import com.google.common.annotations.VisibleForTesting;
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.Value;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.Writer;
+import java.net.InetAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketException;
+import java.net.URI;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/** Opens a minimal HTTP server that retrieves {@literal code} and {@literal state} from the browser. */
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+@RequiredArgsConstructor
+@Slf4j
+public class OAuthValuesReceiver {
+
+  private static final Pattern CODE_PATTERN = Pattern.compile("code=([^ &]+)");
+  private static final Pattern STATE_PATTERN = Pattern.compile("state=([^ &]+)");
+  private static final List<String> ALLOWED_HOSTS = List.of("localhost");
+
+  @VisibleForTesting
+  CompletableFuture<Integer> listenPort = new CompletableFuture<>();
+
+  private final ClientProperties clientProperties;
+  private final PlatformService platformService;
+  private final UserService userService;
+  private final I18n i18n;
+
+  public CompletableFuture<Values> receiveValues(Optional<URI> redirectUri, Optional<OAuthEndpoint> oAuthEndpoint) {
+    return CompletableFuture.supplyAsync(() -> {
+      Collection<URI> redirectUris = getRedirectUris(redirectUri, oAuthEndpoint);
+
+      for (URI uri : redirectUris) {
+        try {
+          return readWithUri(uri);
+        } catch (SocketException e) {
+          log.warn("Port " + uri.getPort() + " is probably already in use", e);
+        }
+      }
+      throw new IllegalStateException("Could not read from any redirect URI: " + redirectUris);
+    });
+  }
+
+  @SneakyThrows
```

Comment:
This is not needed as nothing throws a checked exception here that I can see

Reply:
Good catch